integration/spadev0/util.go

```diff
@@ -0,0 +1,197 @@
+		env.GetString(env.LotusAPIUrl, "https://api.node.glif.io/rpc/v0"),
+		env.GetString(env.LotusAPIToken, ""),
+		providerCacheTTL)
+	if err != nil {
+		panic(err)
+	}
+
+	// Check public IP address
+	ipInfo, err := resolver.GetPublicIPInfo(ctx, "", "")
+	if err != nil {
+		panic(err)
+	}
+	logger.With("ipinfo", ipInfo).Infof("Public IP info retrieved")
+
+	// For each SPID, assemble retrieval tasks for it
+	for spid, replicas := range replicasToTest {
+		strSpid := fmt.Sprintf("f0%d", spid)
+		t, r := prepareTasksForSP(ctx, requester, strSpid, ipInfo, replicas, locationResolver, *providerResolver)
+
+		tasks = append(tasks, t)
+		results = append(results, r)
+	}
+
+	// Write resulting tasks and results to the DB
+	taskClient, err := mongo.
+		Connect(ctx, options.Client().ApplyURI(env.GetRequiredString(env.QueueMongoURI)))
+	if err != nil {
+		panic(err)
+	}
+	taskCollection := taskClient.
+		Database(env.GetRequiredString(env.QueueMongoDatabase)).Collection("task_queue")
+
+	if len(tasks) > 0 {
+		_, err = taskCollection.InsertMany(ctx, tasks)
+		if err != nil {
+			return errors.Wrap(err, "failed to insert tasks")
+		}
+	}
+
+	resultClient, err := mongo.Connect(ctx, options.Client().ApplyURI(env.GetRequiredString(env.ResultMongoURI)))
+	if err != nil {
+		panic(err)
+	}
+	resultCollection := resultClient.
+		Database(env.GetRequiredString(env.ResultMongoDatabase)).
+		Collection("task_result")
+
+	if len(results) > 0 {
+		_, err = resultCollection.InsertMany(ctx, results)
+		if err != nil {
+			return errors.Wrap(err, "failed to insert results")
+		}
+	}
+
+	return nil
+}
+
+var spadev0Metadata map[string]string = map[string]string{
+	"retrieve_type": "spade",
+	"retrieve_size": "1048576",
+	// todo: specify # of cids to test per layer of the tree TBD
+}
+
+func prepareTasksForSP(
+	ctx context.Context,
+	requester string,
+	spid string,
+	ipInfo resolver.IPInfo,
+	replicas []Replica,
+	locationResolver resolver.LocationResolver,
+	providerResolver resolver.ProviderResolver,
+) (tasks []interface{}, results []interface{}) {
+
+	providerInfo, err := providerResolver.ResolveProvider(ctx, spid)
+	if err != nil {
+		logger.With("provider", spid).
+			Error("failed to resolve provider")
+		return
+	}
+
+	location, err := locationResolver.ResolveMultiaddrsBytes(ctx, providerInfo.Multiaddrs)
+	if err != nil {
+		if errors.As(err, &requesterror.BogonIPError{}) ||
+			errors.As(err, &requesterror.InvalidIPError{}) ||
+			errors.As(err, &requesterror.HostLookupError{}) ||
+			errors.As(err, &requesterror.NoValidMultiAddrError{}) {
+
+			results = addErrorResults(requester, ipInfo, results, spid, providerInfo, location,
+				task.NoValidMultiAddrs, err.Error())
+		} else {
+			logger.With("provider", spid, "err", err).
+				Error("failed to resolve provider location")
+			return
+		}
+	}
+
+	for _, document := range replicas {
+		tasks = append(tasks, task.Task{
+			Requester: requester,
+			Module:    task.HTTP, // TODO: Bitswap
```

Comment:
@xinaxu , do you think we should have an entirely separate Module here? ex `task.Spade`? Reason I ask is I don't see the `metadata` `retrieve_type` used anywhere (maybe I'm not looking in the right place though).. is is OK to use that metadata for control flow?

Reply:
plus, with the 3-level tree retrieval method we need, will need to modify both the `bitswap` worker and the `http` worker. Maybe easier to do that in a single Spade worker to leave those other ones untouched (maybe we can still call them from the Spade module)

Reply:
Finally, have you checked the logic here. This is where retrieval worker polls the task queue for tasks. If we end up filling the task but no-one is pulling from it, the task queue will get full and no new tasks will be added. I think you're fine by looking at your logic but want to make sure that you're aware.
https://github.com/data-preservation-programs/RetrievalBot/blob/342f3508af872e610f786919e3bf6fdfa101a1db/pkg/task/task_worker.go#L98

Reply:
> Meanwhile, please double check the current retrieval logic to make sure that it will not panic or crash on unexpected metadata/retrieve_type. You should be fine just by looking at your code

Yup, confirmed - the metadata `retrieve_type` is not actually read anywhere in the codebase, so should not cause any issues.

> Finally, have you checked the logic here. This is where retrieval worker polls the task queue for tasks. If we end up filling the task but no-one is pulling from it, the task queue will get full and no new tasks will be added. I think you're fine by looking at your logic but want to make sure that you're aware.

should we add some kind of back-pressure check to decide to not run the integration if the queue is too full?